File: backend/app/api/v1/fonts.py
# SPDX-License-Identifier: AGPL-3.0-or-later
from typing import Annotated

from fastapi import APIRouter, Depends, Query

from app.core.fonts import preview_file_url, search_fonts
from app.core.security import get_current_user
from app.models.user import User
from app.schemas.font import FontSearchResult

router = APIRouter(prefix="/fonts", tags=["fonts"])


@router.get("/search", response_model=list[FontSearchResult])
async def search_font_catalog(
    current_user: Annotated[User, Depends(get_current_user)],
    q: str,
    limit: Annotated[int, Query(ge=1, le=10)] = 10,
) -> list[FontSearchResult]:
    matches = await search_fonts(q, limit)
    return [
        FontSearchResult(
            id=entry.id,
            family=entry.family,
            category=entry.category,
            variable=entry.variable,
            preview_url=preview_file_url(entry),
        )
        for entry in matches
    ]

File: frontend/packages/api-client/src/api/fonts.ts
// SPDX-License-Identifier: AGPL-3.0-or-later
import { API } from "@api-client/constants";
import { FontSearchResult } from "@api-client/models";
import { getJson } from "@api-client/utils.ts";

const baseUrl = `${API.BASE_URL}/fonts`;

/** Top `limit` (backend-capped at 10) Fontsource catalog matches for `q` - empty for a blank query. */
export function searchFonts(q: string, limit?: number): Promise<FontSearchResult[]> {
  const params = new URLSearchParams({ q });
  if (limit !== undefined) params.set("limit", String(limit));

  return getJson(`${baseUrl}/search?${params}`);
}
